src/Files.App/Actions/FileSystem/FlattenToRootAction.cs

```diff
@@ -0,0 +1,126 @@
+﻿// Copyright (c) 2024 Files Community
+// Licensed under the MIT License. See the LICENSE.
+
+using Microsoft.Extensions.Logging;
+using System.IO;
+using Windows.Storage;
+
+namespace Files.App.Actions
+{
+	internal sealed class FlattenToRootAction : ObservableObject, IAction
+	{
+		private readonly IContentPageContext context;
+		private readonly IGeneralSettingsService GeneralSettingsService = Ioc.Default.GetRequiredService<IGeneralSettingsService>();
+
+		public string Label
+			=> "FlattenToRoot".GetLocalizedResource();
+
+		public string Description
+			=> "FlattenToRootDescription".GetLocalizedResource();
+
+		public bool IsExecutable =>
+			GeneralSettingsService.ShowFlattenOptions &&
+			context.ShellPage is not null &&
+			context.HasSelection &&
+			context.SelectedItem?.PrimaryItemAttribute is StorageItemTypes.Folder;
+
+		public FlattenToRootAction()
+		{
+			context = Ioc.Default.GetRequiredService<IContentPageContext>();
+
+			context.PropertyChanged += Context_PropertyChanged;
+			GeneralSettingsService.PropertyChanged += GeneralSettingsService_PropertyChanged;
+		}
+
+		public Task ExecuteAsync(object? parameter = null)
+		{
+			if (context.ShellPage?.ShellViewModel is null)
+				return Task.CompletedTask;
+
+			var items = context.SelectedItems;
+
+			if (items is null || !items.Any() || items.Any(item => !item.IsFolder))
+				return Task.CompletedTask;
+
+			foreach (var item in items)
+				FlattenFolderAsync(item.ItemPath);
+
+			return Task.CompletedTask;
+		}
+
+		private Task FlattenFolderAsync(string folderPath)
```

Comment:
@0x5bfa where do you recommend moving this code to?

Reply:
This method is not an async method and is not meant to return Task. The return value should be void.

Reply:
I think here is good.
Do you have something else in your mind?

Reply:
I thought we might want to move it to a service or helper.